### PR TITLE
[FLINK-15118][FLINK-15119][FLINK-15120] Make flink-scala-shell use Executors

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

This PR makes the `scala-shell` use the newly introduced `Executors` instead of the old `PlanExecutor`. The changelog mainly includes the partial rewriting of the `FlinkShell` and that of the `ScalaShell(Stream)Environment`.

## Verifying this change

This change is covered by existing tests and it was also verified by manually running examples on Yarn.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 1]
```diff
@@ -1686,5 +1671,23 @@ private static boolean isUsrLibDirIncludedInShipFiles(List<File> shipFiles) {
 			.map(File::getName)
 			.noneMatch(name -> name.equals(DEFAULT_FLINK_USR_LIB_DIR));
 	}
+
+	private void setClusterEntrypointInfoToConfig(final ApplicationReport report) {
```
Comment (TisonKun) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: Nice change.

File: flink-scala-shell/src/main/java/org/apache/flink/api/java/ScalaShellEnvironment.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	private static Configuration validateAndGetConfiguration(final Configuration configuration) {
+		if (!ExecutionEnvironment.areExplicitEnvironmentsAllowed()) {
+			throw new InvalidProgramException(
+					"The RemoteEnvironment cannot be instantiated when running in a pre-defined context " +
+							"(such as Command Line Client, Scala Shell, or TestEnvironment)");
+		}
+		return checkNotNull(configuration);
+	}
+
+	private static List<URL> getJarFiles(final String[] jars) {
```
Comment (TisonKun) [inline flink-scala-shell/src/main/java/org/apache/flink/api/java/ScalaShellEnvironment.java]: Is this pattern the same as that in streaming env? I can see it as a general `JarUtils` function; shall we move it there?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/environment/RemoteStreamExecutionEnvironmentTest.java [hunk 3]
```diff
@@ -57,9 +53,6 @@
 /**
  * Tests for the {@link RemoteStreamEnvironment}.
  */
-@RunWith(PowerMockRunner.class)
```
Comment (TisonKun) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/environment/RemoteStreamExecutionEnvironmentTest.java]: If so, can we remove power mockito all from this test?

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java (hunk 1 above)
Comment (aljoscha) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java]: Indeed! ☕

File: flink-scala-shell/src/main/java/org/apache/flink/api/java/ScalaShellEnvironment.java (hunk 2 above)
Comment (aljoscha) [reply flink-scala-shell/src/main/java/org/apache/flink/api/java/ScalaShellEnvironment.java]: I will do that. 👌

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/environment/RemoteStreamExecutionEnvironmentTest.java (hunk 3 above)
Comment (aljoscha) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/environment/RemoteStreamExecutionEnvironmentTest.java]: Done!

**Comment by TisonKun** [comment]
Although I'm trying to reuse `TestingClusterClient` it is in test scope and we can hardly get access of it. Compiling the patch now and try to run some sample jobs.

**Comment by kl0u** [comment]
@TisonKun thanks for the testing. We also tested it manually on a Yarn cluster and there was no problem. Our tests involved simple graphs similar to `WordCount`. Further testing is always welcome so if @zjffdu can give it go it would be much appreciated. In other case, there will be further testing during the testing period coming after the feature freeze.

File: flink-scala-shell/src/main/scala/org/apache/flink/api/scala/FlinkShell.scala [hunk 4]
```diff
@@ -221,41 +185,59 @@ object FlinkShell {
...
     }
 
     println(" good bye ..")
   }
 
-  def deployNewYarnCluster(
-      configuration: Configuration,
-      configurationDirectory: String,
-      yarnConfig: YarnConfig) = {
+  private def fetchConnectionInfo(config: Config, flinkConfig: Configuration) = {
```
Comment (zjffdu) [inline flink-scala-shell/src/main/scala/org/apache/flink/api/scala/FlinkShell.scala]: It's better to specify the return type explicitly for readability.

**Comment by kl0u** [comment]
Thanks for the reviews! Merged.

**Pull request closed without merging**